### PR TITLE
DOCS: Document logging out of CILogon

### PR DESCRIPTION
**Pull request by GeorgianaElena**
Does @2i2c-org/tech-team agree that adding this note to the docs is acceptable for fixing https://github.com/2i2c-org/infrastructure/issues/957? 

Note [this comment in the original issue](https://github.com/2i2c-org/infrastructure/issues/957#issuecomment-1028036063) that gives more details about it.

### CODE REVIEW

**Comment by GeorgianaElena** [comment]
>Can we maybe rephrase this along the lines of "If a user mistakenly provides an incorrect email when logging in, they will need to... before doing...". Or maybe "If a user wishes to change which email address they login with,...". I'm having trouble keeping track between the "first" and "second" attempts.

Sure! There's a bit of a mess in my head right now :exploding_head: 

Two thoughts:

1. I'm a bit reserved using `email address` to refer to the user account.

So, with CILogon, the login process looks like this:

Hub login page :arrow_right: Auth0 Login screen presenting the CILogon option :arrow_right: CILogon IDP selection page :arrow_right: **IDP login page** :arrow_right: hub access

If on the screen in bold, the same identity provider is used, let's say Google, then a logout must happen before using another email account. But if one chooses the berkeley Identity Provider, then they will be able to login with their berkeley email address without logging out. 

2. If this phrase `If a user mistakenly provides an incorrect email when logging in` means that the incorrect email resolved to a `403` error then this is what happens https://github.com/2i2c-org/infrastructure/issues/973 and a CILogon logout won't solve things.

Hope I didn't create more confusion :sweat:

**Comment by sgibson91** [comment]
> 1. I'm a bit reserved using `email address` to refer to the user account.

Ok, let's definitely say something general like "account information"

> If this phrase `If a user mistakenly provides an incorrect email when logging in` means that the incorrect email resolved to a 403 error then this is what happens https://github.com/2i2c-org/infrastructure/issues/973 and a CILogon logout won't solve things.

Ah, so do we have to do something in the hub at this point? (I will check the issue in case you've answered it there)

**Review by sgibson91 — COMMENTED** [review]
I like [this](https://github.com/2i2c-org/infrastructure/pull/974/commits/c562d67b9fc82aedcf93ceb48a8fcb7c8a343dc2), I feel it's much easier to follow. Thank you! 🙌🏻

**Comment by GeorgianaElena** [comment]
Thanks @sgibson91! I tried to rephrase the note a bit. LTMK if I managed to be more clear and shorter than in the [novel above](https://github.com/2i2c-org/infrastructure/pull/974#issuecomment-1028996147) :sweat:

**Comment by GeorgianaElena** [comment]
You already answered :heart: Thank you!!!!

**Comment by choldgraf** [comment]
So just to make sure I understand, there are two "things to log out of":

**If a user logs into the hub with CILogon, but they use the wrong institutional address**. Then they'll need to log out of two things:

- First, log out of JupyterHub so that they don't keep hitting 403 errors. To do this, go to `https://{hub-name}/hub/logout`
- Second, log out of CILogon so that they're asked to select a new institution. To do this, go to `https://cilogon.org/logout`

When that is done, they will be able to "start fresh" and choose a new CILogon provider and/or account name.

Is that right? If so, I think we should also document this in our user docs as well, since I suspect this will happen to people.

**Comment by damianavila** [comment]
> Is that right? If so, I think we should also document this in our user docs as well, since I suspect this will happen to people.

I agree with this thought.

**Comment by GeorgianaElena** [comment]
> If a user logs into the hub with CILogon, but they use the wrong institutional address. Then they'll need to log out of two things:
> First, log out of JupyterHub so that they don't keep hitting 403 errors. To do this, go to https://{hub-name}/hub/logout

Yep. Checkout https://github.com/2i2c-org/pilot-homepage/pull/8 for a possible "documentation" of this step

> Second, log out of CILogon so that they're asked to select a new institution. To do this, go to https://cilogon.org/logout

Almost. So, they're asked to select an institution regardless of they go to https://cilogon.org/logout. But if they want to use the same institution, but just another address, then in order to be asked which institutional address to use, they will need to go to https://cilogon.org/logout.

> Is that right? If so, I think we should also document this in our user docs as well, since I suspect this will happen to people.

What do you think if we go with the button in https://github.com/2i2c-org/pilot-homepage/pull/8 for the first one and the note in this PR for the second one? Do you think we need to add more details in other places of the docs?

**Comment by choldgraf** [comment]
I took a pass at the language to try and clarify things a little bit in these docs - however I think we will need to iterate a little bit because the instructions didn't quite work for me, so I likely got something wrong :-)

Could you take a look at my latest commit and let me know what I messed up? I think that your strategy sounds good to me as long as we can document the full "how to log out" process here as a start. In the future if the button isn't enough and people are still confused, we can make an addition to the user docs

File: docs/howto/configure/auth-management.md [hunk 1]
```diff
@@ -232,3 +223,26 @@ config:
...
+
+### Switching user accounts or institutions
+
+By default, logging in with a particular user account will persist your credentials in future sessions.
+This means that you'll automatically re-use the same institutional and user account if you try to log back in.
+
+**To switch user accounts**, a user can go to the URL endpoing `https://{hub-name}/hub/logout`.
+The next time they go to the hub's landing page, they'll be asked to re-authenticate.
+
+**To switch CILogon institutions**, a user must go to the [CILogon logout page](https://cilogon.org/logout) and click the button to log out of their institutional account. When they try to log back in they should be directed to a page to select institutions once again.
```
Comment (GeorgianaElena) [inline docs/howto/configure/auth-management.md]: I believe these are actually the other way around. In practice I've havn't seen switching institution require logging out of CILogon.

**Review by GeorgianaElena — COMMENTED** [review]
Thanks for clarifying the docs @choldgraf. I left a couple of comments with my thoughts, but also added [a commit](https://github.com/2i2c-org/infrastructure/pull/974/commits/7d48028325bfaa971a0d81ac44df5f6ec55f6e10) to modify things a bit again. Sorry for not providing suggestions directly, but I wanted to add images too to make things more clear. LTMK what you think and if they work

**Comment by choldgraf** [comment]
Ah thanks for those updates @GeorgianaElena - I think they are good! I pushed a commit with some minor cleanups and clarifications as well. What do you think?

However, following the instructions didn't quite work for me. Here are the steps I followed and what happened. Can you help me debug?

1. Logged-in to demo.pilot.2i2c.cloud with Berkeley credentials
2. 403 page
2. Tap button to log out
3. Take me back to Auth0 login page
4. Choose CILogon
5. Straight to Berkeley log-in, cannot choose Identity Provider
6. Tried going to https://cilogon.org/logout/ and logging out
7. Same thing happened again

I would have expected that step 4 would have then asked me to **choose an identity provider**, but this didn't happen. Can you clarify?

**Comment by GeorgianaElena** [comment]
> 6. Straight to Berkeley log-in, cannot choose Identity Provider

I was able to reproduce your workflow if I check the `remember` box. Did you checked the remember box here?
remeber

If so, then probably you need to clear the cookies. It doesn't have to do with the logout process, but it's probably worth mentioning in the docs here.

> ! I pushed a commit with some minor cleanups and clarifications as well. What do you think?

I don't see any new commit from you @choldgraf :eyes: Maybe you forgot to push it?
https://github.com/2i2c-org/infrastructure/pull/974/commits

**Comment by GeorgianaElena** [comment]
I clarified the docs a bit more and mentioned this scenario when clearing the cookies is required

**Comment by choldgraf** [comment]
@GeorgianaElena the cookie fix worked! Thanks for clarifying, and sorry that I forgot to push my commit before! I've just pushed another commit to clean up the language a bit. I added `sphinx-copybutton` while I was at it :-)

File: docs/howto/configure/auth-management.md [hunk 2]
```diff
@@ -175,31 +175,33 @@ Presently, this involves a few more manual steps than the `auth0` setup describe
...
-Some notions about CILogon authentication worth mentioning:
+### Key terms
 
-1. **Identity Provider**.
-   This is the authentication service available through the CILogon connection.
+Some key terms about CILogon authentication worth mentioning:
 
-   When a user logs in via CILogon, they are first presented with a list of various institutions and organization that they may choose from (e.g. `UC Berkeley` or `Australia National University`).
+Identity Provider
+: The authentication service available through the CILogon connection.
```
Comment (GeorgianaElena) [inline docs/howto/configure/auth-management.md:183]: Had no idea about this syntax, but super cool :+1:

**Review by choldgraf — APPROVED** [review]
This looks good to me now!

**Comment by choldgraf** [comment]
OK I am merging this one in! I think these are nice changes and we can continue updating them as we use CILogon more and learn!
